src/libraries/Common/src/System/Net/Http/aspnetcore/Http2/Hpack/Huffman.cs

```diff
@@ -307,128 +402,145 @@ public static (uint encoded, int bitLength) Encode(int data)
-            // This is done with the following logic:
-            //
-            // if we're at the first entry in the table,
-            //    codeMax = the # of symbols encoded in b bits
-            // else,
-            //    left-shift codeMax by the difference between b and the previous entry's bit length,
-            //    then increment codeMax by the # of symbols encoded in b bits
-            //
-            // Next, we look at the value v encoded in the highest b bits of data. If v is less than codeMax,
-            // those bits correspond to a Huffman encoded symbol. We find the corresponding decoded
-            // symbol in the list of values associated with bit length b in the decoding table by indexing it
-            // with codeMax - v.
+                // Check for correct EOS, which is padding with ones till end of byte
+                // when we STARTED new huffman code in last 8 bits (lookupTableIndex was reset to 0 -> root lookup table).
+                if (lookupTableIndex == 0)
+                {
+                    // Check if all remaining bits are ones.
+                    uint ones = uint.MaxValue >> (32 - bitsInAcc);
+                    if ((acc & ones) == ones)
+                    {
+                        // Is it a EOS. See: http://httpwg.org/specs/rfc7541.html#rfc.section.5.2
+                        break;
+                    }
+                }
 
-            int codeMax = 0;
+                // Lookup index has to be 8 bits aligned to MSB
+                lookupIndex = (byte)(acc << (8 - bitsInAcc));
 
-            for (int i = 0; i < _decodingTable.Length && _decodingTable[i].codeLength <= validBits; i++)
-            {
-                (int codeLength, int[] codes) = _decodingTable[i];
+                ushort lookupValue = decodingTree[(lookupTableIndex << 8) + lookupIndex];
 
-                if (i > 0)
+                if (lookupValue < 0x80_00)
                 {
-                    codeMax <<= codeLength - _decodingTable[i - 1].codeLength;
-                }
+                    // Octet found.
+                    // +---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+
+                    // | 0 |     number_of_used_bits   |              octet            |
+                    // +---+---------------------------+-------------------------------+
+                    bitsInAcc -= lookupValue >> 8;
 
-                codeMax += codes.Length;
+                    if (bitsInAcc < 0)
+                    {
+                        // Last looked up code had more bits than was left in accumulator which indicated invalid or incomplete source
+                        throw new HuffmanDecodingException(SR.net_http_hpack_huffman_decode_failed);
+                    }
 
-                int mask = int.MinValue >> (codeLength - 1);
-                long masked = (data & mask) >> (32 - codeLength);
+                    if (j >= dst.Length)
```

Comment:
nit: looks like `j` should never be > `dst.Length`. can make this a little clearer by changing comparison.

```suggestion
                    if (j == dst.Length)
```